content/en/blog/2023/ambient-ebpf-redirection/index.md

```diff
@@ -0,0 +1,72 @@
+publishdate: 2023-03-15
+attribution: "Iris Ding (Intel), Chun Li (Intel)"
+keywords: [istio,ambient,ztunnel,eBPF]
+---
+
+In Istio ambient mode, the istio-cni component running on each Kubernetes worker node is responsible for redirecting application pod traffic to ztunnel on that node. By default it relies on iptables and
+[Geneve](https://www.rfc-editor.org/rfc/rfc8926.html) tunnels to achieve this redirection. Now, a new approach which is based on eBPF is also available in Istio for this purpose.
+
+## Why eBPF
+
+Although performance considerations are essential in the implementation of Istio ambient mode redirection, it's also important to consider ease of programmability, to enable the implementation of versatile and customized requirements. With eBPF, you can leverage additional context in the kernel to bypass complex routing and simply send packets to there final destination. Furthermore,
+eBPF enables deeper visibility and additional context for packets in the kernel, allowing for more efficient and flexible management of data flow compared with iptables.
+
+## How it works
+
+An eBPF program, attached to the [traffic control](https://man7.org/linux/man-pages/man8/tc-bpf.8.html) ingress and egress hook, has been compiled into istio-cni component. The istio-cni component will watch pod events and attach/detach the eBPF program to other related network interfaces when the pod is moved into or out of ambient mode.
+
+{{< image width="55%"
+    link="ambient-ebpf.png"
+    caption="ambient eBPF architecture"
+    >}}
+
+Using an eBPF program (instead of iptables) eliminates the need to encapsulate tasks (for Geneve), allowing the routing tasks to be customized in the kernel space instead. This yields gains in both performance and flexibility in routing.
+
+To summarize, all traffic from/to the application pod will be intercepted by eBPF and redirected to the corresponding ztunnel pod. On the ztunnel side, proper redirection will be performed based on connection lookup results within the eBPF program. This provides a more efficient control over the network traffic between the application and ztunnel.
+
+## How to enable eBPF redirection in Istio ambient mode
+
+Follow the [Getting Started with Ambient Mesh](/blog/2022/get-started-ambient/) instructions to set up your cluster but, when you install Istio, set the `values.cni.ambient.redirectMode` configuration parameter to `ebpf`:
+
+{{< text bash >}}
+$ istioctl install --set profile=ambient --set values.cni.ambient.redirectMode="ebpf"
+{{< /text >}}
+
+Check the istio-cni logs to confirm eBPF redirection is on:
+
+{{< text plain >}}
+ambient Writing ambient config: {"ztunnelReady":true,"redirectMode":"eBPF"}
+{{< /text >}}
+
+## Performance gains
+
+The latency and throughput (QPS) for eBPF mode is somewhat better than IPtables mode. The following tests were run in a kind cluster with

```

Comment:
I followed what I thought was thsame steps and saw
RPS@256 connections:
ebpf: 114k
iptables: 248k

RPS@16 connections:
ebpf: 42k
iptables: 126k

latency @ 1000qps 16c:
ebpf: .6ms
iptables: .35ms


Seems pretty contrary to the results seen here?

This was running on a single node kind cluster on top of a 48 core GCE machine

Reply:
oof, I think there is a bug causing the redirection to be lost entirely, so "iptables" was actually "no ztunnel"

Reply:
https://github.com/istio/istio/issues/43991 :slightly_smiling_face: 

Real numbers:
RPS@256 connections:
ebpf: 114k
iptables: 118k

RPS@16 connections:
ebpf: 42k
iptables: 44k

latency @ 1000qps 16c:
ebpf: .6ms
iptables: .58ms


So I still see a slight edge for iptables

Reply:
Weird...  My test result is base on istio(5905cf2a037835f3c4ba490f685ed712aa9593e0) and ztunnel(fbb8bf6d2ee015675e73458eb2fc99f02d3af41a).
And I disabled ebpf debug log with my test by:
```
diff --git a/cni/pkg/ebpf/app/ambient_redirect.bpf.c b/cni/pkg/ebpf/app/ambient_redirect.bpf.c
index ff42a6f8b6..fe7711fc73 100644
--- a/cni/pkg/ebpf/app/ambient_redirect.bpf.c
+++ b/cni/pkg/ebpf/app/ambient_redirect.bpf.c
@@ -18,10 +18,12 @@
 
 #include "ambient_redirect.h"
 
-#define dbg(fmt, ...)                               \
-    ({                                              \
-        bpf_printk("[Debug]: " fmt, ##__VA_ARGS__); \
-    })
+//#define dbg(fmt, ...)                               \
+//    ({                                              \
+//        bpf_printk("[Debug]: " fmt, ##__VA_ARGS__); \
+//    })
+
+#define dbg(fmt, ...)  
 
 
 /* This is an array to store host network(ip) info
```
I retested on my env, even with ebpf debug log on, it has a slight edge for iptables.

Reply:
If someone needs to fork Istio to get good performance then that means no one will see this. Please benchmark what users will *actually get* when they follow the steps, not if they run some undocumented forked code.

Reply:
BTW, which is OS & kernel version of your GCE machine?

Reply:
6.1.15, its ~Debian

Reply:
@howardjohn https://github.com/istio/istio/pull/44044  ebpf debug log is turned off by default